llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1587,3 +1588,24 @@ void VPlanTransforms::createInterleaveGroups(
       }
   }
 }
+
+void VPlanTransforms::licm(VPlan &Plan) {
+  VPRegionBlock *LoopRegion = Plan.getVectorLoopRegion();
+  VPBasicBlock *Preheader =
+      cast<VPBasicBlock>(LoopRegion->getSinglePredecessor());
+  // Hoist any loop invariant recipes from the vector loop region to the
+  // preheader.
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+           vp_depth_first_shallow(LoopRegion->getEntry()))) {
+    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
+      // TODO: Relax checks in the future, e.g. we could also hoist reads, if
+      // their memory location is not modified in the vector loop.
+      if (R.mayHaveSideEffects() || R.mayReadFromMemory() || R.isPhi() ||
+          any_of(R.operands(), [](VPValue *Op) {
```

Comment:
I'd recommend changing this to if `if (all_of(R.operands(), ... Op->isDefinedOutsideVectorRegions())) R.moveBefore(...)`, and having the continue branch for `R.mayHaveSideEffects()` etc.

Reply:
Also, I was thinking: the right condition for LICM is "def dominates all its uses"; do you think this is feasible to check with the current infrastructure?

Reply:
I was commenting purely from the perspective of ease of comprehension of the negative-`any_of`, but it's not a strong preference.